### PR TITLE
Prompt tuning for Summarizer AI

### PR DESCRIPTION
**Pull request by LorenzoMinto**
Resolves https://github.com/brave/brave-browser/issues/30171

Prompt should conform to the following format:
```
\n\nHuman: [HUMAN_TEXT]\n\nAssistant: [ASSISTANT_TEXT]\n\nHuman: [MORE HUMAN TEXT]\n\nAssistant:
```

This PR addresses the following issues:
- nits with prompt formatting
- summaries and responses are too long and filled with excessive detail
- instructions-following (if simply added at the beginning) wanes after just a few interactions, and agent goes free-form 

## Submitter Checklist:

- [ ] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/resources/ai_chat_prompts.grdp [hunk 1]
```diff
@@ -1,7 +1,26 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <grit-part>
-  <message name="IDS_AI_CHAT_SUMMARIZE_PROMPT" translateable="false" desc="Default prompt for summarization">
-    Write a concise summary and always respond in bullet points:
-    <ph name="TEXT">$1</ph>
-  </message>
+  <message name="IDS_AI_CHAT_SUMMARIZE_PROMPT" translateable="false" desc="Default prompt for summarization">Here is an article in &lt;article&gt; tags:
+&lt;article&gt;
+<ph name="ARTICLE">$1</ph>
+&lt;/article&gt;
```
Comment (petemill) [inline components/resources/ai_chat_prompts.grdp:6]: Should we strip any string "</article>" and "<article>" form the article text?

File: components/ai_chat/ai_chat_tab_helper.cc [hunk 2]
```diff
@@ -288,24 +289,30 @@ void AIChatTabHelper::CleanUp() {
...
-  AddToConversationHistory(turn);
+  std::string prompt = base::ReplaceStringPlaceholders(
+      l10n_util::GetStringUTF8(IDS_AI_CHAT_SUMMARIZE_PROMPT),
+      {article_text_, GetConversationHistoryString(), turn.text}, nullptr);
 
   std::string prompt_with_history =
-      base::StrCat({GetConversationHistoryString(), ai_chat::kAIPrompt});
+      base::StrCat({prompt, ai_chat::kAIPrompt, " <response>\n"});
+
+  if (turn.visibility != ConversationTurnVisibility::DONT_ADD) {
```
Comment (petemill) [inline components/ai_chat/ai_chat_tab_helper.cc:299]: Did you forget to commit the .mojom file? This branch doesn't have the `DONT_ADD` member.

**Review by petemill — COMMENTED** [review]
It's interesting
- Uses lots of characters to get the job done, maybe we can optimize
- It does stay in character - responds to "David" (eventually) and uses the Sorry, I didn't understand that phrase quite a bit
- Negative: The summary it generated was far too short on https://www.theverge.com/23711172/google-amp-accelerated-mobile-pages-search-publishers-lawsuit I think we should try to give it maybe a percentage or the sentences in the article, or say "up to" a larger number
- Can probably still break out, so not sure what we're gaining with the <article> but it does seem to help it's prompt.
- Why do we need the end <response> tags?

File: components/ai_chat/ai_chat_tab_helper.cc (hunk 2 above)
Comment (LorenzoMinto) [reply components/ai_chat/ai_chat_tab_helper.cc:299]: forgot to push 👍

File: components/resources/ai_chat_prompts.grdp (hunk 1 above)
Comment (LorenzoMinto) [reply components/resources/ai_chat_prompts.grdp:6]: good point. We should probably do 👍

**Comment by petemill** [comment]
> By "breaking out" here are you referring to the article content
I'm referring to the issue where the end of the article might sound like a command and then the AI would answer it independently. But I'm hoping your further instructions at the end of the prompt help there.

**Comment by petemill** [comment]
Thanks for the answers @LorenzoMinto. I think this is worth trying and iterating further from here as a base. Are you ok to add the tag stripping from the article content or would you like help with that? I'm guessing the name "David" is a placeholder, and we can substitute after internal discussion.

**Comment by LorenzoMinto** [comment]
added article tags stripping, let me know how it looks. I also increased the maximum sentence number and this seems to leave the llm more room to decide the number of necessary sentences. As for David, sad to see him go, but yeah it was just a placeholder. We could call the agent Brave AI.

**Review by nullhook — CHANGES_REQUESTED** [review]

File: components/ai_chat/ai_chat_tab_helper.h [hunk 3]
```diff
@@ -77,6 +77,7 @@ class AIChatTabHelper : public content::WebContentsObserver,
 
   // TODO(nullhook): Abstract the data model
   std::vector<ai_chat::mojom::ConversationTurn> chat_history_;
+  std::string article_text_;
```
[the remark this answers is not in the document]
Comment (LorenzoMinto) [reply components/ai_chat/ai_chat_tab_helper.h:80]: because we need this to build the prompt on each api call. Should I be saving it some other way?

File: components/ai_chat/ai_chat.mojom [hunk 4]
```diff
@@ -11,6 +11,7 @@ enum CharacterType {
 };
 
 enum ConversationTurnVisibility {
+  DONT_ADD,
```
[the remark this answers is not in the document]
Comment (nullhook) [reply components/ai_chat/ai_chat.mojom:14]: If it makes sense to consider all three values together as different levels of visibility or inclusion in the chat history, then it may make sense to keep them in the same enum. However, if DONT_ADD represents a completely different concept that is not related to visibility or chat history, it may be clearer to have a separate enum or variable for it.

I have plans to eventually abstract the data model to a separate interface and have the following variants for visibility:

1. Private: Only visible to the system and added to the history
2. Public: Visible to the user and added to the history

I guess you can keep it as DONT_ADD or INTERNAL for now and maybe add a comment. I'll clean it up in a follow-up PR. As per this PR change looks like HIDDEN variant is not being used anywhere.

`DONT_ADD, // does not get added to chat history, used as a flag internally`

File: components/ai_chat/ai_chat_tab_helper.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (nullhook) [reply components/ai_chat/ai_chat_tab_helper.h:80]: If we're no longer attaching the summarized text, then subsequent API calls will include the entire article text, which increases the message length and could impact on cost. What is the benefit of this approach and how does it affect the answers from Claude on subsequent runs?

[the remark this answers is not in the document]
Comment (LorenzoMinto) [reply components/ai_chat/ai_chat_tab_helper.h:80]: I was under the impression that we were always adding the article text (and the summarized text, which we could remove as redundant). We add the initial prompt to the conversation, right? Also, if we were not, the user would be asking follow up questions on the summarized text, which has been stripped of most details. 

PS: just checked on the ai_chat logs for latest master and we do include the article on each call as the start of the conversation history. Am I missing something?

[the remark this answers is not in the document]
Comment (LorenzoMinto) [reply components/ai_chat/ai_chat_tab_helper.h:80]: I agree an interesting optimisation would be to produce a long summary and base questions on that, but it would have to be a pretty comprehensive summary, which we might not want to show to the user for UX reasons. But I think we can look at optimisations in a future PR.

File: components/ai_chat/ai_chat.mojom (hunk 4 above)
[the remark this answers is not in the document]
Comment (LorenzoMinto) [reply components/ai_chat/ai_chat.mojom]: Yeah, a separate enum with `Type` might be better then. I changed it to INTERNAL, I liked it better. Then you can handle the decoupling in the refactor you mentioned 👍

File: components/ai_chat/ai_chat_tab_helper.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (nullhook) [reply components/ai_chat/ai_chat_tab_helper.h:80]: Okay, so now I understand where the change is. Previously, we used to store the article text along with the prompt in the conversation history itself. However, with the current approach, we no longer store it in the history directly. Instead, we keep the article text separate and attach it along with the prompt on each API call.

The separation of article text in-memory feels like it's adding complexity, requiring additional handling to ensure association with the conversation history on each API call. However, I see a benefit of this is on subsequent API calls. Perhaps a feature: "User-controlled summarization level" where users can switch between summary lengths or toggle between the summarized version and the full article text based on their preferences.

I'm curious if @petemill has thoughts on this separation. If not, then I'll go ahead and approve this PR.

[the remark this answers is not in the document]
Comment (nullhook) [reply components/ai_chat/ai_chat_tab_helper.h:80]: I almost forgot to mention this: make sure to clear the article text in the CleanUp function. You will also notice other members being cleared there.

[the remark this answers is not in the document]
Comment (LorenzoMinto) [reply components/ai_chat/ai_chat_tab_helper.h:80]: done 👍

File: browser/ui/webui/ai_chat/ai_chat_ui_page_handler.cc [hunk 5]
```diff
@@ -53,6 +53,8 @@ void AIChatUIPageHandler::SubmitHumanConversationEntry(
   // TODO(nullhook): Abstract prompt injection cleanups to a central place
   base::ReplaceSubstringsAfterOffset(&input_copy, 0, ai_chat::kHumanPrompt, "");
   base::ReplaceSubstringsAfterOffset(&input_copy, 0, ai_chat::kAIPrompt, "");
+  base::ReplaceSubstringsAfterOffset(&input_copy, 0, "<article>", "");
+  base::ReplaceSubstringsAfterOffset(&input_copy, 0, "</article>", "");
```
Comment (petemill) [inline browser/ui/webui/ai_chat/ai_chat_ui_page_handler.cc:57]: Should we strip history and question tags from the input rather than article?

File: components/ai_chat/ai_chat_tab_helper.cc [hunk 6]
```diff
@@ -261,24 +262,27 @@ void AIChatTabHelper::DistillViaAlgorithm(const ui::AXTree& tree) {
   base::ReplaceSubstringsAfterOffset(&contents_text, 0, ai_chat::kHumanPrompt,
                                      "");
   base::ReplaceSubstringsAfterOffset(&contents_text, 0, ai_chat::kAIPrompt, "");
+  base::ReplaceSubstringsAfterOffset(&contents_text, 0, "<article>", "");
+  base::ReplaceSubstringsAfterOffset(&contents_text, 0, "</article>", "");
```
Comment (petemill) [inline components/ai_chat/ai_chat_tab_helper.cc:266]: Do you think we need to worry about white space and variations, e.g. `</ article x>`

File: browser/ui/webui/ai_chat/ai_chat_ui_page_handler.cc (hunk 5 above)
Comment (LorenzoMinto) [reply browser/ui/webui/ai_chat/ai_chat_ui_page_handler.cc:57]: True. Should we sanitise out all _privileged_ tags for both article and subsequent outputs?

File: components/ai_chat/ai_chat_tab_helper.cc (hunk 6 above)
Comment (LorenzoMinto) [reply components/ai_chat/ai_chat_tab_helper.cc:266]: It really depends on their tokenizer. Checking with the chatGPT tokenizer, these are the token ids translations:
```
<article>Something</article>     ------> [27, 20205, 29, 22210, 3556, 20205, 29]
<article >Something</ article>   ------> [27, 20205, 1875, 22210, 3556, 2708, 29]
```
So the space do seem to make a difference. If that's the case we don't need to worry about variations.

Comment (petemill) [reply components/ai_chat/ai_chat_tab_helper.cc:265]: You're saying think this won't confuse the API because it will be very strict if the tokens are different?

```
Here is an article in &lt;article&gt; tags:
<article>
This is my article content about xml issues like </ article> and other things and <article real=true>
</article>

You will be acting as a news expert named David created by the company Brave News. Your goal is to summarize and answer questions about the news article above to users in a way that is simple and time efficient. You will be replying to users who are on the Brave browser and who will be confused if you don't respond in the character of David.
Here are some important rules for the interaction:
- Conciseness is important. Your responses should not exceed 6 sentences.
- Always respond in a neutral tone.
- Always stay in character, as David, an AI from Brave News.
- If you are unsure how to respond, say "Sorry, I didn't understand that. Could you rephrase your question?"

Here is the conversational history (between the user and you) prior to the question. It could be empty if there is no history:
<history>
</history>

Here is the user's question:
<question>
Summarize the above article
</question>

How do you respond to the user's question? Put your response in &lt;response&gt;&lt;/response&gt; tags.
``````

**Review by petemill — APPROVED** [review]
This is great. There are more things to consider and iterate on but since this is behind a flag we should get it in for internal feedback.

**Review by nullhook — APPROVED** [review]
lgtm
